src/Stratis.Features.FederatedPeg/Distribution/DistributionManager.cs

```diff
@@ -0,0 +1,138 @@
+        // It is therefore not in any miner's advantage to delay or skip producing their blocks as it will affect their proportion of the produced blocks.
+        // We pay no attention to whether a miner has been kicked since the last distribution or not.
+        // If they produced an accepted block, they get their reward.
+
+        public DistributionManager(Network network, ChainIndexer chainIndexer, ILoggerFactory loggerFactory)
+        {
+            this.network = network;
+            this.chainIndexer = chainIndexer;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            this.epoch = this.network.Consensus.MaxReorgLength == 0 ? DefaultEpoch : (int)this.network.Consensus.MaxReorgLength;
+            this.lastDistributionHeight = 0;
+        }
+
+        private int GetDistributionEpochStart(int blockHeight)
+        {
+            // This is a special case which will not be the case on the live network.
+            if (blockHeight < this.epoch)
+            {
+                return 0;
+            }
+
+            return blockHeight - (blockHeight % this.epoch) - this.epoch;
+        }
+
+        private int GetDistributionEpochEnd(int blockHeight)
+        {
+            return blockHeight - (blockHeight % this.epoch);
+        }
+
+        /// <inheritdoc />
+        public List<Recipient> Distribute(int mainChainHeight, Money totalReward)
+        {
+            ChainedHeader tip = this.chainIndexer.Tip;
+
+            // We need to determine which block on the sidechain contains a commitment to the height of the mainchain block that originated the reward transfer.
+            // We otherwise do not have a common reference point from which to compute the epoch.
+            // Look back at most 2x epoch lengths to avoid searching from genesis every time.
+
+            // To avoid miners trying to disrupt the chain by committing to the same height in multiple blocks, we loop forwards and use the first occurrence
+            // of a commitment with height >= the search height.
+
+            int blockHeight = 0;
+
+            ChainedHeader currentHeader = this.chainIndexer.Height > (2 * this.epoch) ? this.chainIndexer.GetHeader(this.chainIndexer.Height - (2 * this.epoch)) : this.chainIndexer.Genesis;
+
+            // Cap the maximum number of iterations.
+            int iterations = this.chainIndexer.Height - currentHeader.Height;
+
+            var encoder = new CollateralHeightCommitmentEncoder(this.logger);
+
+            for (int i = 0; i < iterations; i++)
+            {
+                int? commitmentHeight = encoder.DecodeCommitmentHeight(currentHeader.Block.Transactions[0]);
+
+                if (commitmentHeight == null)
+                    continue;
+
+                if (commitmentHeight >= mainChainHeight)
+                {
+                    blockHeight = commitmentHeight.Value;
```

Comment:
Cause then here, `blockheight` will effectively always be the mainchain height? Unless I'm understanding this wrong?

Reply:
Basically this method is running from the sidechain, which has no knowledge of the mainchain's height other than what is committed to in the coinbase transactions (an extra OP_RETURN output of a specific format).

So mainChainHeight is the height the reward transaction was made on the mainchain side, and we look in the sidechain blocks to see what the earliest block is that has committed to that mainchain height or higher.